aiida_common_workflows/workflows/relax_and_bands.py

```diff
@@ -0,0 +1,324 @@
+            inputs[key] = val.value
+        if isinstance(val, orm.Dict):
+            inputs[key] = val.get_dict()
+        if isinstance(val, orm.List):
+            inputs[key] = val.get_list()
+        if isinstance(val, orm.Code):
+            inputs[key] = val.label
+        if isinstance(val, dict):
+            deserialize(val)
+
+    return inputs
+
+
+@calcfunction
+def seekpath_explicit_kp_path(structure, seekpath_params):
+    """
+    Return the modified structure of SeekPath and the explicit list of kpoints.
+    :param structure: StructureData containing the structure information.
+    :param seekpath_params: Dict of seekpath parameters to be unwrapped as arguments of `get_explicit_kpoints_path`.
+    """
+    from aiida.tools import get_explicit_kpoints_path
+
+    results = get_explicit_kpoints_path(structure, **seekpath_params)
+
+    return {'structure': results['primitive_structure'], 'kpoints': results['explicit_kpoints']}
+
+
+def validate_inputs(value, _):  #pylint: disable=too-many-branches,too-many-return-statements
+    """Validate the entire input namespace."""
+
+    process_class = WorkflowFactory(value['relax_sub_process_class'].value)
+    generator = process_class.get_input_generator()
+
+    # Validate that the provided ``relax_inputs`` are valid for the associated input generator.
+    try:
+        generator.get_builder(**AttributeDict(value['relax_inputs']))
+    except Exception as exc:  # pylint: disable=broad-except
+        return f'`{generator.__class__.__name__}.get_builder()` fails for the provided `relax_inputs`: {exc}'
+
+    #Validate that the plugin for bands and the relax are the same
+    bands_plugin = value['bands_sub_process_class'].value.replace('common_workflows.bands.', '')
+    relax_plugin = value['relax_sub_process_class'].value.replace('common_workflows.relax.', '')
+    if relax_plugin != bands_plugin:
+        return 'Different code between relax and bands. Not supported yet.'
+
+
+def validate_sub_process_class_r(value, _):
+    """Validate the sub process class."""
+    try:
+        process_class = WorkflowFactory(value.value)
+    except exceptions.EntryPointError:
+        return f'`{value.value}` is not a valid or registered workflow entry point.'
+
+    if not inspect.isclass(process_class) or not issubclass(process_class, CommonRelaxWorkChain):
+        return f'`{value.value}` is not a subclass of the `CommonRelaxWorkChain` common workflow.'
+
+
+def validate_sub_process_class_b(value, _):
+    """Validate the sub process class."""
+    try:
+        process_class = WorkflowFactory(value.value)
+    except exceptions.EntryPointError:
+        return f'`{value.value}` is not a valid or registered workflow entry point.'
+
+    if not inspect.isclass(process_class) or not issubclass(process_class, CommonBandsWorkChain):
+        return f'`{value.value}` is not a subclass of the `CommonBandsWorkChain` common workflow.'
```

Comment:
Instead of having these two separate functions that are almost identical, I would add the following validator as a classmethod to the workchain
```python

    @staticmethod
    def validate_sub_process_class(value, _, required_class=None):
        """Validate the sub process class."""
        try:
            process_class = WorkflowFactory(value)
        except exceptions.EntryPointError as exc:
            return f'`{value}` is not a valid or registered workflow entry point: {exc}'

        if not inspect.isclass(process_class) or not issubclass(process_class, required_class):
            return f'`{value}` is not a subclass of the `{required_class}` common workflow.'
```
Then you can specify the ports for the sub process classes as follows:
```python

        spec.input(
            'sub_process_class_relax',
            non_db=True,
            required=False,
            validator=partial(cls.validate_sub_process_class, required_class=CommonRelaxWorkChain)
        )
        spec.input(
            'sub_process_class_scf',
            non_db=True,
            required=False,
            validator=partial(cls.validate_sub_process_class, required_class=CommonRelaxWorkChain)
        )
        spec.input(
            'sub_process_class_bands',
            non_db=True,
            validator=partial(cls.validate_sub_process_class, required_class=CommonBandsWorkChain)
        )
```